src/main/java/TcpServer.java

```diff
@@ -0,0 +1,128 @@
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.util.Optional;
+import java.util.concurrent.*;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+public class TcpServer implements Closeable {
+  // TODO: make logger instance-specific, prefixing logs w/ config & object info
+  static final Logger logger = Logger.getLogger(TcpServer.class.getName());
+
+  // configuration used to initialize the serverSocket
+  final ServerConfig config;
+
+  // number of clients currently connected to the server socket
+  private final AtomicInteger connectionCount;
+
+  // executor for handling of new connections
+  private final ExecutorService connectionHandlerExecutor;
+
+  // Thread where new connections are accepted from serverSocket
+  private Optional<Thread> acceptThread;
+
+  // socket which manages client connections
+  // Initialized to empty, and set to a Socket object once started.
+  private Optional<ServerSocket> serverSocket;
+
+  TcpServer(ServerConfig config) {
+    this.config = config;
+    this.serverSocket = Optional.empty();
+    this.acceptThread = Optional.empty();
+    this.connectionCount = new AtomicInteger(0);
+    this.connectionHandlerExecutor = Executors.newFixedThreadPool(this.config.maxConnections);
+  }
+
+  /**
+   * @see connectionCount
+    */
+  int getConnectionCount() {
+    return connectionCount.get();
+  }
+
+  /**
+   * Start listening on the specified port.
+   *
+   * Must not be called more than once.
+   *
+   * @throws IOException If the underlying socket couldn't bind successfully.
+   */
+  void start() throws IOException {
+    assert !serverSocket.isPresent();
+    ServerSocket socket = new ServerSocket(this.config.port, this.config.maxConnections);
+    this.serverSocket = Optional.of(socket);
+    Thread acceptThread = new Thread(this::acceptConnections);
+    this.acceptThread = Optional.of(acceptThread);
+    acceptThread.start();
+  }
+
+  // Accept and handle new connections
+  private void acceptConnections() {
+    assert acceptThread.isPresent();
+    assert Thread.currentThread().equals(acceptThread.get());
+    assert serverSocket.isPresent();
+
+    ServerSocket serverSocket = this.serverSocket.get();
```

Comment:
Here and in a couple of other places made me wonder to what extent you'd see simplification and improved unit testability if you were to pass in your `ServerSocket` or an abstraction / wrapper version. I suspect that would give you a seam where you could pass a fake that didn't need to do network communication, and you could force the code down all these exception paths from unit tests.

Reply:
Perhaps it could be handled internally that way, via classes like `SocketProvider` and `ConnectionHandler`. However, I think the consumer API of creating a server with a config, starting it, and stopping it, should remain the same.

My intuition is probably weaker than yours in this regard, so I would probably give this class some room to grow before refactoring it too invasively. For example, when implementing the actual connection handling & response.

Reply:
Yeah, there's often a balance between client convenience and flexibility. We can often kind of cheat, by starting on the flexibility side, for both testing and production-code flexibility reasons, and then adding something like a [Facade](https://sourcemaking.com/design_patterns/facade) to hide all the configuration & provide the client convenience.